dynamiqs/integrators/mcsolve/diffrax_integrator.py

```diff
@@ -0,0 +1,258 @@
+        time_diffs = jnp.where(time_diffs < 0, jnp.inf, time_diffs)
+        save_index = jnp.argmin(time_diffs)
+        # allocate memory for the state that is updated during the loop
+        save_state = SaveState(
+            ts=first_result.ts[0],
+            ys=jax.vmap(self.save)(unit(first_result.ys[0])),
+            save_index=save_index,
+            jump_times=jnp.full(self.options.max_jumps, jnp.nan),
+        )
+        state = State(
+            save_state=save_state,
+            final_state=first_result.ys[1][0, :, :],
+            final_time=first_jump_time,
+            num_jumps=0,
+            key=sample_key,
+        )
+
+        def _outer_cond_fun(_state):
+            return _state.final_time < self.t1
+
+        def _outer_body_fun(_state):
+            jump_time = _state.final_time
+            psi = _state.final_state
+            _next_key, _rand_key, _sample_key = jax.random.split(_state.key, num=3)
+            jump_op = self._sample_jump_ops(jump_time, psi, _sample_key)
+            psi_after_jump = unit(jump_op @ psi)
+            # Construct a new linspace over the remaining times with the same shape as
+            # self.ts (because the shape can't be dynamic!). Note that jax does not
+            # support dynamic array sizes: this makes vmaps of loops tricky as discussed
+            # here https://github.com/patrick-kidger/equinox/issues/870. The issue is
+            # that some trajectories terminate before others do (they experience fewer
+            # jumps), but the loop body needs to keep executing until the last
+            # trajectory finishes. In these "idling" cases, new_times is a constant
+            # array with fill value self.t1. The interpolator
+            # dx.backward_hermite_coefficients below requires the time inputs to be
+            # monotonically strictly increasing, which would then throw an error. So in
+            # these cases we pass a placeholder of self.ts.
+            new_times = jnp.linspace(jump_time, self.t1, len(self.ts))
+            new_times = jnp.where(_state.final_time < self.t1, new_times, self.ts)
+            # This new random value should be uniform over [0, 1), not restricted to
+            # [no_jump_prob, 1) like in the first case.
+            _rand = jax.random.uniform(_rand_key)
+            result_after_jump = self._run(psi_after_jump, new_times, _rand)
+            # extract saved y values to interpolate, replace infs with nans
+            psis_after_jump = result_after_jump.ys[0]
+            psi_inf_to_nan = jtu.tree_map(
+                lambda _y: jnp.where(jnp.isinf(_y), jnp.nan, _y), psis_after_jump
+            )
+            # setting fill_forward_nans_at_end prevents nan states from getting saved
+            psi_coeffs = dx.backward_hermite_coefficients(
+                new_times, psi_inf_to_nan, fill_forward_nans_at_end=True
+            )
+            psi_interpolator = dx.CubicInterpolation(new_times, psi_coeffs)
```

Comment:
This is a very tricky point and something that significantly complicates the implementation. The issue is mentioned in the first two lines of the extended comment, which is that `new_times` needs to have the same shape as `self.ts`. It is not enough for the shape of the output of `_outer_body_fun` to be static, but indeed all array sizes need to be determined at compile time for all code underneath a jit. Thus the `new_times` array must have the same shape as `self.ts`. 

What I don't think I've explored that this comment made me think of (maybe this is what you meant) is making `new_times` an array that has a static shape, filled only with times from the original `self.ts` and then padded with nans. Preliminary tests with `dx.diffeqsolve` indicate tha diffrax does not complain if some of the times in `SaveAt` are nans. This would avoid the need for interpolation, I think.

Reply:
See 2cf82b6d1725555e05f35577a0470ddc2d01fb91 below for a simplified implementation that does not require interpolation based on the above discussion. Thanks very much for making this point!